File: 트리/leetcode652(Medium)-Find Duplicate Subtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# DFS + 백트래킹 : 53ms(70.76%), 24.5MB(23.66%)
class Solution:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        visited = {}
        result = set()

        def dfs(node):
            if not node:
                return "()"

            left = dfs(node.left)
            right = dfs(node.right)

            representation = "(" + left + str(node.val) + right + ")"

            if representation in visited:
                result.add(representation)
            else:
                visited[representation] = node
            
            return representation

        dfs(root)

        return [visited[representation] for representation in list(result)]

File: 트리/leetcode652(Medium)-Find Duplicate Subtrees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

// DFS + 백트래킹 : 97ms(90.68%), 58.4MB(27.97%)
/**
 * @param {TreeNode} root
 * @return {TreeNode[]}
 */
var findDuplicateSubtrees = function (root) {
  const visited = {}
  const result = new Set()

  const dfs = (node) => {
    if (!node) {
      return "()"
    }

    const left = dfs(node.left)
    const right = dfs(node.right)

    const representation = `(${left}${node.val}${right})`

    if (representation in visited) {
      result.add(representation)
    } else {
      visited[representation] = node
    }

    return representation
  }

  dfs(root)

  return Array.from(result).map((representation) => visited[representation])
}
